### DISCUSSION TITLE
ConfigurableResource with Generic type

### DISCUSSION THREAD

**Discussion by stefan-adelbert-lassio** [dagster-io/dagster#22083] [Q&A]
I'd like to create a `ConfigurableResource` but use a generic type for its member. For example,

```python
from typing import TypeVar, Generic
import pydantic
import dagster

class BasicCredentials(pydantic.BaseModel):
    username: str
    password: str

CredentialsT = TypeVar('CredentialsT', bound=pydantic.BaseModel)

class Credentials(dagster.ConfigurableResource, Generic[CredentialsT]):
    _credentials: CredentialsT = pydantic.PrivateAttr()

    @property
    def credentials(self) -> CredentialsT:
        return self._credentials

    def setup_for_execution(self, context: dagster.InitResourceContext) -> None:
        dummy_secret = {"username": "Me", "password": "MyPassword"}
        self._credentials = CredentialsT.model_validate(dummy_secret)

credentials = Credentials[BasicCredentials]()
```

I'm getting this error,
```
TypeError: All parameters must be present on typing.Generic; you should inherit from typing.Generic[~TResValue, ~CredentialsT].
```

It is possible to do this kind of thing with pydantic directly - see https://docs.pydantic.dev/latest/concepts/models/#generic-models.

Could someone give me some pointers on getting this approach working with a `dagster.ConfigurableResource`?

**Comment by stefan-adelbert-lassio** [comment]
I have come up with an alternative approach which uses a factory function rather than generics. 

```python
from typing import TypeVar, Generic
import pydantic
import dagster

CredentialsT = TypeVar('CredentialsT', bound=pydantic.BaseModel)

def make_credentials_resource(credentials_type: Type[CredentialsT]) -> dagster.ConfigurableResource:
    class _CredentialsResource(dagster.ConfigurableResource):
        _credentials: CredentialsT = pydantic.PrivateAttr()

        @property
        def credentials(self) -> CredentialsT:
            return self._credentials
        
        def setup_for_execution(self, context: dagster.InitResourceContext) -> None:
            # In reality, this secret would come from a secrets store.
            dummy_secret = {"username": "Me", "password": "MyPassword"}
            self._credentials = credentials_type.model_validate(dummy_secret)
    return _CredentialsResource

class BasicCredentials(pydantic.BaseModel):
    username: str
    password: str

BasicCredentialsResource = make_credentials_resource(BasicCredentials)
basic_credentials = BasicCredentialsResource()

with dagster.build_resources(resources={"credentials": basic_credentials}) as resources:
    print(resources.credentials.credentials)
```

**Reply by benpankow** [reply]
Hi Stefan - I think this should work as long as the credentials type is treated as an inner resource, but I don't think it will be able to properly make any nested config available in the Launchpad or other parts of the Dagster UI. For this, you may be able to use `pydantic.create_model` so that concrete type annotations are available to Dagster when the resource type is first created.

```python
from typing import Any, Type

from dagster import Config, ConfigurableResource, Definitions, ResourceParam, asset
from pydantic import Field, create_model


def make_credentials_resource(
    credentials_type: Type[Config],
) -> ConfigurableResource:
    model = create_model(
        "Credentials",
        credentials=(credentials_type, Field(description="credentials")),
        __base__=ConfigurableResource,
    )

    return model


class BasicCredentials(Config):
    username: str
    password: str


BasicCredentialsResource = make_credentials_resource(BasicCredentials)
basic_credentials = BasicCredentialsResource.configure_at_launch()


@asset
def my_asset(basic_credentials: ResourceParam[Any]): ...


defs = Definitions(
    resources={"basic_credentials": basic_credentials}, assets=[my_asset]
)
```

<img width="436" alt="Screenshot 2024-05-28 at 11 30 17 AM" src="https://github.com/dagster-io/dagster/assets/10215173/5679908f-80df-4efe-b82c-888512318480">

**Reply by stefan-adelbert-lassio** [reply]
@benpankow Thank you for this response!

This approach doesn't work in my case because in my actual use case, the CredentialsResource depends on another ConfigurableResource which provides the raw credentials (from a secrets manager). I left this out in my original post to keep things simpler.

This means that I would need to be able to express that resource dependency in the `create_model` call. Actually, maybe that is indeed possible. I will try that now...

**Reply by stefan-adelbert-lassio** [reply]
@benpankow I tried your approach. Thank you again for the response.

In this case, the credentials should remain internal to the resource and not be exposed in dagster UI, so I'm going to stick with my factory function approach. I can see how your approach could be very useful in other cases though.
